Issue #1332 - Backport promise-based media playback.

athenian200 opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=1244768

I managed to get loom videos working for myself using this patch. I used the version originally submitted to Bugzilla rather than the one that went into mozilla-central, because the final version that went into mozilla-central had a ton of audio-wrapper and media-block refactors. I only had to fix their newer event return syntax in the patches back to the way we use to get it to compile. I couldn't figure out how to get around some of the minor refactoring in the patch itself related to bringing in promises as a concept here, but I managed to avoid the rest of the changes.

I can't guarantee I didn't break anything else in the process of implementing this. I tried asking around to see if it was worth submitting a pull request or not, but the thing is... no one else seems to understand DOM well enough to know what it would take to get this ready for submission. 

For instance, do I need to break down the changes into multiple commits even though it's just three files (because the changes are kind of involved), do more research on the spec until I can explain each part of the change, find some additional test cases, etc?

JustOff commented:
Unfortunately this failed to build on Windows:

```
 7:35.39 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3483): error C2065: 'rv': undeclared identifier
 7:35.39 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3484): error C2065: 'rv': undeclared identifier
 7:35.40 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3484): error C2228: left of '.Failed' must have class/struct/union
 7:35.40 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3484): note: type is 'unknown-type'
 7:35.40 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3485): error C2065: 'rv': undeclared identifier
 7:35.40 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3485): error C2228: left of '.StealNSResult' must have class/struct/union
 7:35.40 d:/Pale-Moon-test/platform/dom/html/HTMLMediaElement.cpp(3485): note: type is 'unknown-type'
```

athenian200 commented:
I actually built this on Windows. I think I fixed that issue already... I uploaded the wrong version because I made the changes to the folder I put under "platform" rather than my working copy of UXP.

It was caused by a typo I made that I corrected already.

JustOff commented:
Thanks, now it builds w/o errors. I can confirm this PR resolves the [Loom player](https://www.loom.com/share/6601f93702f3486aa2fe9f314f8b17e9) issue, and I also did a brief check playing media on YouTube, Vimeo and several other sites - everything seems fine.

Alas, I can't add anything about the backport itself, as I have very little experience in this field.

JustOff commented:
I can also confirm that this PR resolves the NRK.no issue reported [here](https://forum.palemoon.org/viewtopic.php?f=3&t=23495).

JustOff commented:
Tag #1332.
